alf/environments/alf_gym3_wrapper.py

```diff
@@ -0,0 +1,415 @@
+
+        Note that by default Gym3 environmenst cannot be reset. However, if
+        ``support_force_reset`` is set to True, action -1 will be sent to all
+        the sub-environments to reset them if the underlying environment follows
+        the convention.
+
+        Otherwise the reset will be ignored.
+
+        """
+        if self._support_force_reset:
+            self._gym3_env.act(np.array([-1] * self.batch_size))
+            self._done = [False] * self.batch_size
+        else:
+            logging.warning('reset() ignored by AlfGym3Wrapper')
+
+        reward, observation, _ = self._gym3_env.observe()
+
+        time_step = self._create_time_step(
+            reward,
+            observation,
+            step_type=[ds.StepType.FIRST] * self.batch_size,
+            # Faking actions
+            action=nest.map_structure(
+                lambda spec: spec.numpy_zeros(outer_dims=(self.batch_size, )),
+                self.action_spec()))
+
+        return time_step
+
+    # Implement abstract _step
+    def _step(self, action) -> ds.TimeStep:
+        """Implement ``_step()`` for compatibility
+
+        A very important note here is that special treatment is done at the end
+        of each episode. Unlike Gym environments, Gym3 environments do not
+        return ``done=True`` immediately. Instead, ``first=True`` is returned
+        upon the ``observe()`` of the NEXT ``act()``, which is actually the
+        FIRST FRAME of a new episode in Gym3 sense.
+
+        With this wrapper we CHANGED that definition, so that the frame with
+        ``first=True`` actually becomes the LAST FRAME of the previous episode.
+        Because the observation of the new episode can be DRAMATICALLY DIFFERENT
+        from the actual last frame of the previous episode, the previous
+        observation is returned in this case.
+
+        Therefore to summarize, effectively we will
+
+        1. Repeat the end-of-episode observation twice for each episode.
+
+        2. Throw away the first frame of each episode, and use the second frame
+           as if it is the first frame.
+
+        Gym3's official Gym wrapper DID THE SAME.
+
+        """
+        _, prev_observation, _ = self._gym3_env.observe()
+
+        np_action = nest.map_structure(lambda x: x.cpu().numpy(), action)
+        self._gym3_env.act(np_action)
+
+        reward, observation, first = self._gym3_env.observe()
+
+        # Override the obervation with the previous observation if that
+        # particular environment has ``first=True``.
+        for i in range(self.batch_size):
+            if first[i]:
+                observation[i] = prev_observation[i]
+
+        # This does the trick of repeating end-of-episode frames and throwing
+        # away first-of-episode frames.
+        step_type = [
+            ds.StepType.FIRST if d else
+            (ds.StepType.LAST if f else ds.StepType.MID)
+            for d, f in zip(self._done, first)
+        ]
```

Comment:
Will this create one-frame misbatch between the rendered frame and policy action when playing a model?

Reply:
Actually I think there might be some problem for value learning if the end-of-episode frame is repeated twice, depending on the reward of a "first" frame. In theory the value network will return the same value given the two identical frames, then the bellman equation won't be satisfied between the last two frames?

Reply:
Right, I thought about this when I see `gym3`'s official handling on this and it does put two identical frames at the end of each episode. My understanding is that when the length of the episode is long enough, the number of end frame is actually small compared to other frames, and hopefully this does not alter the training too much.

> timeout. The same value network will be called twice on the same frame to get V, for example. Then we have a TD learning loss between the last two identical frames: (V + r - \gamma V)^2 which will never be zero.

Totally agree with the analysis here.

Reply:
I added a `TODO` to 

1. acknowledge the issue here
2. as a reminder to more properly fix it in the future.